### PR TITLE
feat: Allow changing persistence

### PR DESCRIPTION
**Pull request by benjackwhite**
## Changes

Fixes https://github.com/PostHog/posthog-js/issues/1024

* Also adds tailwind to the next playground

Allows doing something like this:

```ts
// Cookieless tracking with swap to cookie-full tracking when enabled
let cookieOptIn = () => localStorage.getItem("cookie-opt-in") === "true"

posthog.init(KEY, { persistence: cookieOptIn() ? "localStorage+cookie" : "memory" }

function optInToCookies() {
    localStorage.setItem("cookie-opt-in", "true")
    posthog.update_config({ persistence: "localStorage+cookie" })
}
```

## Follow up
- [ ] Update docs and notify team to update tutorials (to answer these questions [1](https://posthog.com/questions/question-8) [2](https://posthog.com/questions/prevent-posthog-storing-cookies-until-user-consents)

## Checklist
- [x] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [x] Accounted for the impact of any changes across different browsers
- [x] Accounted for backwards compatibility of any changes (no breaking changes in posthog-js!)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "author": "hey@posthog.com",
     "license": "MIT",
     "scripts": {
-        "start": "pnpm build-rollup -w",
+        "start": "pnpm build-react && pnpm build-rollup -w",
```
Comment (pauldambra) [inline package.json:9]: silly question maybe... this now builds react once and then builds rollup and starts watching just the rollup build

what's the benefit of building react here?

File: playground/nextjs/src/CookieBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        posthog.set_config({ persistence: consentGiven ? 'localStorage+cookie' : 'memory' })
+    }, [consentGiven])
+
+    if (consentGiven === null) return null
+
+    return (
+        <div className="fixed right-2 bottom-2 border rounded p-2 bg-gray-100 text-sm">
+            {!consentGiven ? (
+                <>
+                    <p>I am a cookie banner - hear me roar.</p>
```
Comment (pauldambra) [inline playground/nextjs/src/CookieBanner.tsx:27]:
```suggestion
                    <p>I am a cookie banner - you wouldn't like m when I'm angry.</p>
```

technically biscuits can't roar... but banners do say that :p

File: src/__tests__/posthog-persistence.test.ts [hunk 3]
```diff
@@ -167,5 +167,40 @@ describe('persistence', () => {
...
+            // Swap to localStorage
+            newConfig = makePostHogConfig('test', 'localStorage')
+            lib.update_config(newConfig, config)
+            config = newConfig
+
+            // Check store contains data and props are the same
+            expect(document.cookie).toEqual('')
+            expect(localStorage.getItem('ph__posthog')).toEqual(JSON.stringify(expectedProps()))
+            expect(lib.properties()).toEqual(expectedProps())
+        })
```
Comment (pauldambra) [inline src/__tests__/posthog-persistence.test.ts:204]: :chef's kiss:

File: src/posthog-persistence.ts [hunk 4]
```diff
@@ -40,24 +53,17 @@ export class PostHogPersistence {
...
-        let token = ''
-
-        if (config['token']) {
-            token = config['token'].replace(/\+/g, 'PL').replace(/\//g, 'SL').replace(/=/g, 'EQ')
-        }
-
+        this.config = config
         this.props = {}
         this.campaign_params_saved = false
-        this.custom_campaign_params = config['custom_campaign_params'] || []
```
Comment (pauldambra) [inline src/posthog-persistence.ts:53]: lost `this.custom_campaign_params = config['custom_campaign_params'] || []` on purpose?

File: src/posthog-persistence.ts [hunk 5]
```diff
@@ -68,28 +74,28 @@ export class PostHogPersistence {
...
+            store = localPlusCookieStore
         } else {
-            this.storage = cookieStore
+            store = cookieStore
         }
 
-        this.load()
-        this.update_config(config)
-        this.save()
+        return store
```
Comment (pauldambra) [inline src/posthog-persistence.ts:98]: nicer that this building and saving is separate 👍

**Review by pauldambra — APPROVED** [review]
i trust you ran it locally and since the new test passes this looks good to me

File: package.json (hunk 1 above)
Comment (benjackwhite) [reply package.json:9]: React needs to exist, but it rarely if ever changes. So this issue was around running the next playground locally, it meant I had to call build first and then start which was simply annoying

Comment (pauldambra) [reply package.json:9]: Ah, I'm a bad person and never use the playground 🙈

File: src/posthog-persistence.ts (hunk 4 above)
Comment (benjackwhite) [reply src/posthog-persistence.ts:53]: We have the whole config now so we just reference it directly where it is used.

File: playground/nextjs/src/CookieBanner.tsx (hunk 2 above)
Comment (benjackwhite) [inline playground/nextjs/src/CookieBanner.tsx:27]:
```suggestion
                    <p>I am a cookie banner - you wouldn't like me when I'm hangry.</p>
```

**Review by mariusandra — APPROVED** [review]
Had a look through the code, but didn't run. What I saw looks good, and seconding Paul: seems like the test passes, so I trust it works.

**Pull request merged**